dev-docs/RFCs/v8.0/image-layer-wms-rfc.md

```diff
@@ -0,0 +1,178 @@
+  getImage({boundingBox, width, height, layers, parameters}): Promise<Image>;
+  getLegendImage({layers, parameters}): Promise<Image>;
+  getFeatureInfo({layers, parameters}): Promise<ImageFeatureInfo>;
+  getLayerInfo({layers, parameters}): Promise<ImageDataSourceLayerInfo>
+}
+
+import {load, LoaderOptions} from '@loaders.gl/core';
+import {WMSCapabilitiesLoader, WMSFeatureInfoLoader} from '@loaders.gl/wms';
+import {ImageLoader} from '@loaders.gl/images';
+
+class WMSDataSource implements ImageDataSource {
+  url: string;
+  loadOptions: LoaderOptions = {};
+
+  constructor({url, loadOptions: LoaderOptions}) {
+    this.url = url;
+    this.loadOptions = loadOptions;
+  }
+
+  getCapabilities(): Promise<ImageDataSourceCapabilities> {
+    const url = this._getUrl({request: 'GetCapabilities', layers, parameters});
+    return load(url, WMSCapabilitiesLoader, this.loadOptions);
+  }
+
+  getImage({boundingBox, width, height, layers: string[], parameters: Record<string, unknown>}): Promise<Image> {
+    const url = this._getUrl({request: 'GetMap', layers, parameters});
+    return load(url, ImageLoader, this.loadOptions);
+  }
+
+  getLegendImage(options: {layers: string[], parameters: Record<string, unknown>}): Promise<Image> {
+    const url = this._getUrl({request: 'GetLegendImage', layers, parameters});
+    return load(url + '?REQUEST=GetCapabilities', WMSCapabilitiesLoader, this.loadOptions);
+  }
+
+  getFeatureInfo({layers: string[], parameters: Record<string, unknown>}): Promise<ImageFeatureInfo> {
+    const url = this._getUrl({request: 'GetFeatureInfo', layers, parameters});
+    return load(url + '?REQUEST=GetCapabilities', ImageLoader, this.loadOptions);
+  }
+
+  getLayerInfo({layers: string[], parameters: Record<string, unknown>}): Promise<ImageDataSourceLayerInfo> {
+    const url = this._getUrl({request: 'GetLayerInfo', layers, parameters})
+    return load(url + '?REQUEST=GetLayerInfo', WMSLayerInfoLoader, this.loadOptions);
+  }
+
+  /**
+   * @note protected, since perhaps getUrl may need to be overridden to handle certain backends?
+   * @note if override is common, maybe add a callback prop?
+   * */
+  protected getUrl(options: {request: string; layers: string[], parameters: Record<string, unknown>}): string {
+    let url = `${this.url}?REQUEST=${options.request}`;
+    if (options.layers.length) {
+      url += `&LAYERS=[${options.layers.join(',')}]`;
+    }
+  }
+}
+```
+
+This might trigger a process to define similar data sources for other services.
+
+```typescript
+interface VectorTileDataSource {
+  getCapabilities(): Promise<IVectorTileDataSourceCapabilities>;
+  getTile({boundingBox, width, height, layers, parameters}): Promise<VectorTile>;
+}
+```
+
+Such a set of data sources would ideally be completely independent of deck.gl,
+and probably best placed in loaders.gl.
+
+
+## Appendix: Potential Future Improvements
+
+### Mosaicing and Client-Side Tiling ideas
```

Comment:
Client-side tiling indeed sounds rather involved, but my intuition is that it would help to get more reliable and less variable performance. Re-fetching the entire viewport feels a bit antiquated in terms of architecture and makes the frontend even more dependent on the performance of the backend. WMS systems can be rather slow to respond I think depending on how much data needs to be loaded to render the requested viewport.

I've never worked with WMS personally but my feeling is that it would seem simpler to request tiles via the TileLayer instead of fetching a new image for the viewport every time the user pans. Whenever I've viewed maps that refetch the entire viewport on every pan, the UX has been low

Reply:
> I've never worked with WMS personally but my feeling is that it would seem simpler to request tiles via the TileLayer instead of fetching a new image for the viewport every time the user pans. Whenever I've viewed maps that refetch the entire viewport on every pan, the UX has been low

I believe I included the link to the standard OpenLayers tiled WMS example that shows how labeling collapses when the image server is unaware that is serving tiles (zoom out and you get multiple labels per state). Also some GPU backends can render server side and visualize using WMS, and doing multiple calls will crater performance and generate exponential cost increases.

That said, it will work well in many cases and we should also make it trivial to connect tile layer to WMS sources.